src/awkward/_util.py

```diff
@@ -798,7 +799,13 @@ def visitor(left, right) -> bool:
             right = right.to_IndexedOptionArray64()
 
         if type(left) is not type(right):
-            return False
+            if not check_regular and (
+                left.is_list and right.is_regular or left.is_regular and right.is_list
+            ):
+                left = left.to_ListOffsetArray64()
+                right = right.to_ListOffsetArray64()
+            else:
+                return False
```

Comment:
We don't use this here, but I wrote it, and then removed the usage. I think it's good to keep for future tests.